### PR TITLE
Snapshot prune feature & snapshot revert improvement

### PR DESCRIPTION
**Pull request by shuo-wu**
https://github.com/longhorn/longhorn/issues/3613
https://github.com/longhorn/longhorn/issues/3748

### CODE REVIEW

**Comment by innobead** [comment]
@shuo-wu Please update the dep because https://github.com/longhorn/sparse-tools/pull/80 has been merged.

File: pkg/controller/revert.go [hunk 1]
```diff
@@ -12,25 +12,56 @@ import (
...
+			}
+			if disk.Removed {
+				return fmt.Errorf("cannot revert to removed snapshot %v", name)
+			}
+			// for c := range disk.Children {
+			// 	if c == head {
+			// 		logrus.Infof("there is no need to do revert since the parent of the current volume head is already snapshot %v", name)
+			// 		return nil
+			// 	}
+			// }
```
Comment (shuo-wu) [inline pkg/controller/revert.go:53]: Should we check that the revert target is the snapshot that is directly behind the volume head?

**Comment by innobead** [comment]
@shuo-wu LGTM, please help resolve the conflicts before merging.

File: pkg/controller/revert.go (hunk 1 above)
Comment (innobead) [reply pkg/controller/revert.go:53]: If the current volume head has data then the volume is reverted to the previous snapshot, we should still create a new empty volume head pointing to the snapshot, right? If yes, we don't need this check.

Comment (shuo-wu) [reply pkg/controller/revert.go]: YES. I just tested this part. The volume will discard the old volume head regardless of the reverting snapshot being the parent.
I have removed this code.

**Comment by shuo-wu** [comment]
2 fixes for the integration failure:
1. Now the replica will reload the snapshot from the backup **full** restore. This guarantees that the snapshot is in the disk chain before the reverting start.
2. Return error if the target snapshot does not exist or is not in the replica disk chain

**Pull request merged**